### PR TITLE
Issues/#1635 repository lifecycle

### PR DESCRIPTION
**Pull request by aschwarte10**
GitHub issue resolved: #1635 

This change cleans the initialization of the FedX Repository lifecycle.

Main features:
- follow semantics of RDF4J Repository API
-- particularly w.r.t to initialization and shutdown
- no static management of federation (=> multiple federations per JVM)
- properly support external RepositoryResolver and FederatedServiceResolver

Note: this change is a huge refactoring touching many classes. In most cases this is due to replacing the static access to all helper structures with instance data from the FederationContext.

I am proposing this change now, because this is basically the last chance for the required cleanup before the federation is released in RDF4J.

For the review it is recommended to go commit wise as there is a logical order. Commits by themselves should be ok to follow (with the explanations from the commit message)

Note: Either as of this PR or as a follow up I am also planning to remove the static Config class and make it a member of the federation context.

### CODE REVIEW

File: tools/federation/src/main/java/org/eclipse/rdf4j/federated/FedX.java [hunk 1]
```diff
@@ -113,16 +127,29 @@ public ValueFactory getValueFactory() {
 	public void initialize() throws SailException {
```
Comment (jeenbroekstra) [inline tools/federation/src/main/java/org/eclipse/rdf4j/federated/FedX.java]: `Sail#initialize` is deprecated, I think you should rename this to `Sail#init`.

File: tools/federation/src/main/java/org/eclipse/rdf4j/federated/FederationManager.java [hunk 2]
```diff
@@ -85,124 +72,27 @@
...
-		this.cache = cache;
-		this.executor = executor;
-		QueryManager.instance = new QueryManager(this, repo); // initialize the singleton query manager
+	public FederationManager() {
+
 	}
 
-	public FedX getFederation() {
-		return federation;
+	public void initialize(FedX federation, FederationContext federationContext) {
```
Comment (jeenbroekstra) [inline tools/federation/src/main/java/org/eclipse/rdf4j/federated/FederationManager.java]: Same here - for the Sail, the Repository, and the RepositoryManager, we renamed `initialize` to `init`. See #552 for details.

File: tools/federation/src/main/java/org/eclipse/rdf4j/federated/QueryManager.java [hunk 3]
```diff
@@ -52,30 +57,43 @@
...
-	protected QueryManager(FederationManager federationManager, Repository repo) {
+	public QueryManager(FederationManager federationManager, Repository repo) {
 		this.federationManager = federationManager;
 		this.repo = repo;
+
+		BigInteger lastQueryId = new BigInteger("0");
+		this.nextQueryID = new AtomicBigInteger(lastQueryId);
+	}
+
+	public void initialize() {
```
Comment (jeenbroekstra) [inline tools/federation/src/main/java/org/eclipse/rdf4j/federated/QueryManager.java]: I realize this one is not overriding any existing methods, but given #552 perhaps we should rename this to `init` as well?

File: tools/federation/src/main/java/org/eclipse/rdf4j/federated/endpoint/EndpointBase.java [hunk 4]
```diff
@@ -145,11 +146,11 @@ public long size() throws RepositoryException {
 	}
 
 	@Override
-	public void initialize() throws RepositoryException {
+	public void initialize(FederationContext federationContext) throws RepositoryException {
```
Comment (jeenbroekstra) [inline tools/federation/src/main/java/org/eclipse/rdf4j/federated/endpoint/EndpointBase.java]: As above

File: tools/federation/src/main/java/org/eclipse/rdf4j/federated/FedX.java (hunk 1 above)
Comment (aschwarte10) [reply tools/federation/src/main/java/org/eclipse/rdf4j/federated/FedX.java]: Good catch. I now make use of AbstractSail functionality and all the inherited base functionality. Required some more smaller adjustments, see last commit

File: tools/federation/src/main/java/org/eclipse/rdf4j/federated/FederationManager.java (hunk 2 above)
Comment (aschwarte10) [reply tools/federation/src/main/java/org/eclipse/rdf4j/federated/FederationManager.java]: Makes sense to align

File: tools/federation/src/main/java/org/eclipse/rdf4j/federated/QueryManager.java (hunk 3 above)
Comment (aschwarte10) [reply tools/federation/src/main/java/org/eclipse/rdf4j/federated/QueryManager.java]: Done

File: tools/federation/src/main/java/org/eclipse/rdf4j/federated/endpoint/EndpointBase.java (hunk 4 above)
Comment (aschwarte10) [reply tools/federation/src/main/java/org/eclipse/rdf4j/federated/endpoint/EndpointBase.java]: Done

**Comment by aschwarte10** [comment]
All review comments addressed. @jeenbroekstra would you please have one more look at the changes?

**Pull request merged**